packages/jaeger-ui/src/model/ddg/DdgEVManager.tsx

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { TEdge } from '@jaegertracing/plexus/lib/types';
+
+import { compareKeys } from './visibility-key';
+
+import { EDdgEdgeKeys, PathElem, DdgVertex, TDdgModel } from './types';
+
+export default class DdgEVManager {
+  private edges: Set<TEdge>;
+  private edgeToFarSideOfEdgePathElems: Map<TEdge, Set<PathElem>>;
+  private farSideOfEdgePathElemsToEdge: Map<Set<PathElem>, TEdge>;
+  private prevVisKey: string;
+  private pathElemToFarSideOfEdgePathElems: Map<PathElem, Set<PathElem>>;
+  private pathElemToVertex: Map<PathElem, DdgVertex>;
+  private vertices: Map<string, DdgVertex>;
+  private visIdxToPathElem: PathElem[];
+
+  constructor({ ddgModel }: { ddgModel: TDdgModel }) {
+    this.edges = new Set();
+    this.edgeToFarSideOfEdgePathElems = new Map();
+    this.farSideOfEdgePathElemsToEdge = new Map();
+    this.pathElemToFarSideOfEdgePathElems = new Map();
+    this.pathElemToVertex = new Map();
+    this.prevVisKey = '';
```

Comment:
it's used by `getEdgesAndVertices` which in turn calls `compareKeys` to determine which edges and vertices need to be added/removed

Reply:
thinking about how to avoid storing this. would the alternative approach in `getEdgesAndVertices` be:
```
getEdgesAndVertices = (codec: string): { edges: TEdge[], vertices: TDdgVertex[] } => {
  const indices = _map(Array.from(this.pathElemToVertex.keys()), 'visibilityIdx');
  const existingCodec = encode(indices); // encode is new name for createKey
  const { shown, hidden } = compareCodecs({ new: codec, old: existingCodec });
  /* ... unchanged ... */
}
```
?

This seems inefficient compared to storing a variable. Another option instead of `compareCodecs` would be:
```
getEdgesAndVertices = (codec: string): { edges: TEdge[], vertices: TDdgVertex[] } => {
  const currIndices = _map(Array.from(this.pathElemToVertex.keys()), 'visibilityIdx');
  const newIndices = decode(codec);
  const { shown, hidden } = compareIndices({ new: newIndices, old: currIndices });
  /* ... unchanged ... */
}
```

Reply:
`codec` has the `encode` and `decode` functions, the resultant object of `decode` is not the `codec`.